classes/services/UserService.inc.php

```diff
@@ -456,4 +467,137 @@ public function canCurrentUserGossip($userId) {
 
 		return true;
 	}
+
+	/**
+	 * Merge two user accounts, including attributed articles and all associated
+	 * information.
+	 *
+	 * @param $user User The user to merge from. This user will no longer exist
+	 *  when this operation is done.
+	 * @param $mergeIntoUser User The user to merge into. This user will contain
+	 *  all data from $user when this operation is done.
+	 */
+	function mergeUsers($user, $mergeIntoUser) {
+		$userId = $user->getId();
+		$mergeIntoUserId = $mergeIntoUser->getId();
+
+		HookRegistry::call('User::mergeUsers', array(&$userId, &$mergeIntoUserId));
+
+		$noteDao = DAORegistry::getDAO('NoteDAO');
+		$notes = $noteDao->getByUserId($userId);
+		while ($note = $notes->next()) {
+			$note->setUserId($mergeIntoUserId);
+			$noteDao->updateObject($note);
+		}
+
+		$editDecisionDao = DAORegistry::getDAO('EditDecisionDAO');
+		$editDecisionDao->transferEditorDecisions($userId, $mergeIntoUserId);
+
+		$reviewAssignmentDao = DAORegistry::getDAO('ReviewAssignmentDAO');
+		foreach ($reviewAssignmentDao->getByUserId($userId) as $reviewAssignment) {
+			$reviewAssignment->setReviewerId($mergeIntoUserId);
+			$reviewAssignmentDao->updateObject($reviewAssignment);
+		}
+
+		$articleEmailLogDao = DAORegistry::getDAO('SubmissionEmailLogDAO');
+		$articleEmailLogDao->changeUser($userId, $mergeIntoUserId);
+		$articleEventLogDao = DAORegistry::getDAO('SubmissionEventLogDAO');
+		$articleEventLogDao->changeUser($userId, $mergeIntoUserId);
+
+		$submissionCommentDao = DAORegistry::getDAO('SubmissionCommentDAO');
+		$submissionComments = $submissionCommentDao->getByUserId($userId);
+
+		while ($submissionComment = $submissionComments->next()) {
+			$submissionComment->setAuthorId($mergeIntoUserId);
+			$submissionCommentDao->updateObject($submissionComment);
+		}
+
+		$accessKeyDao = DAORegistry::getDAO('AccessKeyDAO');
+		$accessKeyDao->transferAccessKeys($userId, $mergeIntoUserId);
+
+		// Transfer old user's individual subscriptions for each journal if new user
+		// does not have a valid individual subscription for a given journal.
+		$individualSubscriptionDao = DAORegistry::getDAO('IndividualSubscriptionDAO');
+		$oldUserSubscriptions = $individualSubscriptionDao->getByUserId($userId);
+
+		while ($oldUserSubscription = $oldUserSubscriptions->next()) {
+			$subscriptionJournalId = $oldUserSubscription->getJournalId();
+			$oldUserValidSubscription = $individualSubscriptionDao->isValidIndividualSubscription($userId, $subscriptionJournalId);
+			if ($oldUserValidSubscription) {
+				// Check if new user has a valid subscription for current journal
+				$newUserSubscription = $individualSubscriptionDao->newUserSubscription($mergeIntoUserId, $subscriptionJournalId);
```

Comment:
Hmmm... does this function newUserSubscription exist? 
Hmmm... Would it be enough to only check isValidIndividualSubscription as in the the next elseif statement?

Reply:
I actually don't know! I moved all of this over from `UserAction::mergeUsers()`. I saw that you fixed an issue with that code, and ported that over into this https://github.com/pkp/pkp-lib/pull/3656/commits/14c0d726468fb414a814a385003eeb2e3011c4fa. But there might be other issues lingering in there.

I did a quick search and didn't see the function. @asmecher do you know if this needs to be updated?

Reply:
ah I see, I think you can then use getByUserId -- I've already forgotten about that fix :-)

Reply:
It looks like that got mistakenly changed (probably by a regexp gone amok while I was refactoring the subscription stuff). It should be `IndividualSubscriptionDAO::getSubscriptionIdByUser`.

Reply:
Oops, that function has been removed (it's still in 2.x). Yes, you can use `getByUserId`, or perhaps `getByUserIdForJournal`.